Fix #181: float32 accuracy issue

willson-chen opened the pull request:
The issue accuracy is caused by casting from float32 to float64 when encoding, which leads to extra decimal appended to the original float. Transformation from float32 to string, and then float64 could avoid this issue.

willson-chen committed:
Fix #181: float32 accuracy issue

wongoo commented:
a unit test is required for it

willson-chen committed:
Fix go fmt failure

wongoo commented on encode.go:
```diff
@@ -112,7 +113,9 @@ func (e *Encoder) Encode(v interface{}) error {
 
 	case float32:
-		e.buffer = encFloat(e.buffer, float64(val))
+		str := strconv.FormatFloat(float64(val), 'f', -1, 32)
```
suggest to add a new func encFloat32 to receive a float32 and encode it, the string converting would degrade performance

wongoo commented:
suggest to add a new func encFloat32 to receive a float32 and encode it, the string converting would degrade performance

willson-chen commented:
I referenced [dubbo-hessian-lite/PR#12](https://github.com/apache/dubbo-hessian-lite/pull/12) to create this patch.

Adding `func encFloat32` is a good choice. And then we should encode the float32 directly indead of casting it to float64. BTW, casting causes the accuracy issue. But I am not sure if it will introduce any compatibility issues.

```
double     ::= 'D' b7 b6 b5 b4 b3 b2 b1 b0   # 64-bit float
           ::= x5b                   # 0.0
           ::= x5c                   # 1.0
           ::= x5d b0                # byte cast to double
                                     #  (-128.0 to 127.0)
           ::= x5e b1 b0             # short cast to double
           ::= x5f b3 b2 b1 b0       # 32-bit float cast to double
```

wongoo commented:
currently hessian2 encoding all float as 64-bit float, while decoding support multiple cases.
 it's worth to do effort to let the encoding follow the protocol.

AlexStocks commented:
> currently hessian2 encoding all float as 64-bit float, while decoding support multiple cases.
> it's worth to do effort to let the encoding follow the protocol.

we can add comment here. In my opinion, we should keep pace with dubbo-hessian-lite.

willson-chen committed:
Add the unit test case for Issue181

willson-chen commented on double_test.go:
```diff
@@ -42,6 +42,28 @@ func TestEncDouble(t *testing.T) {
 	t.Logf("decode(%v) = %v, %v\n", v, res, err)
 }
 
+func TestIssue181(t *testing.T) {
+	var (
+		v   float32
+		err error
+		e   *Encoder
+		d   *Decoder
+		res interface{}
+	)
+
+	e = NewEncoder()
+	v = 99.8
+	e.Encode(v)
+	if len(e.Buffer()) == 0 {
+		t.Fail()
+	}
+
+	// res would be '99.800003' without patches in PR #196
+	d = NewDecoder(e.Buffer())
+	res, err = d.Decode()
+	t.Logf("decode(%v) = %v, %v\n", v, res, err)
+}
+
```
I followed the previous test case `func TestEncDouble` to add the test case for Issue181. But I am puzzled by that only one judgement on `Encode()` and no judgement on if the result is right.

wongoo replied:
yes, u r right, it should add a judgement for the result too, pls add for it

wongoo commented:
pls check whether dubbo-hessian-lite support 32-bit format `x5f b3 b2 b1 b0       # 32-bit float cast to double`

willson-chen commented:
> pls check whether dubbo-hessian-lite support 32-bit format `x5f b3 b2 b1 b0 # 32-bit float cast to double`

Hi, here is [the implement of float32 encode in dubbo-hessian-lite](
https://github.com/apache/dubbo-hessian-lite/blob/master/src/main/java/com/alibaba/com/caucho/hessian/io/BasicSerializer.java#L108). As we can see, dubbo-hessian-lite cast float to double. So I don't think it supports the 32-bit format.

wongoo commented:
it support exactly, see: https://github.com/apache/dubbo-hessian-lite/blob/aea62f0a8e82cad885cc0f7f568e9c27ef6ea70c/src/main/java/com/alibaba/com/caucho/hessian/io/Hessian2Output.java#L733

willson-chen commented:
@wongoo yeah, you are right. I had a careful check the code block, and foud dubbo-hessian-lite both support `hessian` and `hessian2`. `hessian` has only `double     ::= 'D' b7 b6 b5 b4 b3 b2 b1 b0` format when encoding float and double.

In fact, dubbo-hessian-lite encode the number depending the value, but not the type, `float(32-bit in java)` or `double(64-bit in java)`. All the float would be transform to double, and when the value is less then 3 digits after the decimal point, the number will be encoded in `x5f b3 b2 b1 b0` format.

So what should we do?
* Follow dubbo-hessian-lite, we should add the value judgement and deside which format to encode in `func encFloat` of `double.go`.
* Encode exactly depend on the type, we should add `func encFloat32`.

wongoo commented:
> Encode exactly depend on the type, we should add func encFloat32.

my opinion is adding a new func for float32.

willson-chen committed:
Add encFloat32 in double.go to encode float32 type

- Call encFloat32 to encode float32 while encoding
- Add unit test case to test float32 encoding

willson-chen commented on double_test.go:
```diff
@@ -42,6 +42,33 @@ func TestEncDouble(t *testing.T) {
+	// res would be '99.800003' without patches in PR #196
+	d = NewDecoder(e.Buffer())
+	res, err = d.Decode()
+	f := res.(float64)
```
`decDouble` returns float64 while decoding float32. Considering the compatibility, I didn't change the return type from float64 to float32.

wongoo commented on double.go:
```diff
@@ -62,6 +62,30 @@ END:
+	}
+
+END:
+	iv := int32(v * 1000)
```
it should check whether it's equal `0.001*iv == v`,  and the unit test is not cover it now, see https://github.com/apache/dubbo-hessian-lite/blob/aea62f0a8e82cad885cc0f7f568e9c27ef6ea70c/src/main/java/com/alibaba/com/caucho/hessian/io/Hessian2Output.java#L778

wongoo replied:
otherwise, it should use strconv to convert it to float64 and encode to 64 bit

wongoo reviewed (commented):
pls add unit test for math.MaxFloat32、math.MaxFloat32-1、math.MaxFloat32+1、math.MaxFloat64、math.MaxFloat64-1

willson-chen committed:
Improve encFloat32 of double.go

wongoo commented:
> The command "go fmt && [[ -z `git status -s` ]]" exited with 1.

pls format the code

willson-chen committed:
Fix git fmt failure

willson-chen commented:
> pls add unit test for math.MaxFloat32、math.MaxFloat32-1、math.MaxFloat32+1、math.MaxFloat64、math.MaxFloat64-1

Fair enough. But I need some time to adjust the test case.

> pls format the code

done

wongoo reviewed (approved):
LGTM

wongoo reviewed (approved):
LGTM

zonghaishang reviewed (changes requested):
hessian2规范 float就是64位的，hessian go不应该用float32去表示它，因为可能溢出不能完全解决问题。

我的建议：
1. 针对hessian库支持，应该用float64去处理，不应该用float32，并且不应先转string，因为float是基础类型
2. 应该在readme告知开发者应该使用float64 

----

Hessian2 float is 64-bit, and Hessian Go should not be represented by float32 because overflow may not solve the problem completely.

my advice: 
1. Hessian Go USES Float64 encoding and decoding instead of converting a String first
2. The user should be explicitly informed in the ReadME of Hessian Go that instead of using Float32, it should be treated with Float64.

fangyincheng reviewed (approved):
LGTM

amirfi95 commented:
> @wongoo yeah, you are right. I had a careful check the code block, and foud dubbo-hessian-lite both support `hessian` and `hessian2`. `hessian` has only `double ::= 'D' b7 b6 b5 b4 b3 b2 b1 b0` format when encoding float and double.
> 
> In fact, dubbo-hessian-lite encode the number depending the value, but not the type, `float(32-bit in java)` or `double(64-bit in java)`. All the float would be transform to double, and when the value is less then 3 digits after the decimal point, the number will be encoded in `x5f b3 b2 b1 b0` format.
> 
> So what should we do?
> 
> * Follow dubbo-hessian-lite, we should add the value judgement and deside which format to encode in `func encFloat` of `double.go`.
> * Encode exactly depend on the type, we should add `func encFloat32`.

Hi could you elaborate why you are decoding a double prefixed with x5f as an integer and divide it by 1000.
The name BC_DOUBLE_MIL implies millis (/100) but the spec says decode it as 4 byte float and not as integer.
I'm struggling to match existing implementations against the spec :/

(spec mentioning compact doubles as float : http://hessian.caucho.com/doc/hessian-serialization.html#anchor14)

>::= x5f b3 b2 b1 b0       # 32-bit float cast to double

@willson-chen willson-chen
